presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java

```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
+            }
+            return converters.build();
+        }
+
+        private List<BlockFieldConverter> createConvertersByName(String columnName, List<RowField> prestoTypeFields, List<parquet.schema.Type> fieldTypes)
+        {
+            int fieldCount = fieldTypes.size();
+            int prestoFieldCount = prestoTypeFields.size();
+            if (fieldCount != prestoFieldCount) {
+                log.info("Schema mismatch, metastore schema for row column %s has %s fields but parquet schema has %s fields",
+                        columnName,
+                        prestoFieldCount,
+                        fieldCount);
+            }
 
-            ImmutableList.Builder<BlockConverter> converters = ImmutableList.builder();
-            for (int i = 0; i < prestoTypeParameters.size(); i++) {
+            Map<String, ObjectPair<Integer, Type>> fieldTypeMap = new HashMap<String, ObjectPair<Integer, Type>>(prestoFieldCount);
+            for (int i = 0, n = prestoTypeFields.size(); i < n; i++) {
+                RowField field = prestoTypeFields.get(i);
+                fieldTypeMap.put(field.getName().get().toLowerCase(), ObjectPair.create(Integer.valueOf(i), field.getType()));
+            }
+
+            boolean outOfOrder = false;
+            ImmutableList.Builder<BlockFieldConverter> builder = ImmutableList.builder();
+            for (int i = 0, prevFieldIndex = -1; i < fieldCount; i++) {
                 parquet.schema.Type fieldType = fieldTypes.get(i);
-                converters.add(createConverter(prestoTypeParameters.get(i), columnName + "." + fieldType.getName(), fieldType, i));
+                ObjectPair<Integer, Type> pair = fieldTypeMap.get(fieldType.getName().toLowerCase());
+                if (pair == null) {
+                    log.info("Schema mismatch, metastore schema for row column %s doesn't have field %s but parquet schema has it",
+                            columnName,
+                            fieldType.getName());
+                    builder.add(ParquetDummyPrimitiveConverter.converter);
+                }
+                else {
+                    int currFieldIndex = pair.getFirst().intValue();
+                    if (prevFieldIndex > currFieldIndex) {
+                        outOfOrder = true;
+                    }
+                    builder.add(createConverter(pair.getSecond(), columnName + "." + fieldType.getName(), fieldType, currFieldIndex, true));
+                    prevFieldIndex = currFieldIndex;
+                }
             }
-            this.converters = converters.build();
+            List<BlockFieldConverter> converters = builder.build();
+            if (outOfOrder) {
+                log.info("Schema mismatch, struct field order for metastore schema for row column %s is different from that of parquet schema",
+                        columnName);
+                ImmutableList.Builder<Type> types = ImmutableList.builder();
```

Comment:
add comment here: when reading parquet files, pass converters and parquet types in parquet schema fields order